File: fila/Fila Dinâmica/Java/Fila Dinâmica Dupla Terminacão/FilaDinamicaDuplaTerminacao.java
public class FilaDinamicaDuplaTerminacao implements DuplamenteEnfileiravel {
	
	private int quantidade;
	private int tamanho;
	private NoDuplo ponteiroInicio;
	private NoDuplo ponteiroFim;
	
	public FilaDinamicaDuplaTerminacao(int tamanho) {
		quantidade = 0;
		this.tamanho = tamanho;
		ponteiroInicio = null;
		ponteiroFim = null;
	}

	public FilaDinamicaDuplaTerminacao() {
		this(10);
	}

	//funciona como o enfileirar de FilaDinamica
	@Override
	public void enfileirarFim(Object dado) {
		if(!estaCheia()) {
			NoDuplo noTemporario = new NoDuplo();
			noTemporario.setDado(dado);
			if (!estaVazia()) {
				ponteiroFim.setProximo(noTemporario);
			} else  {
				ponteiroInicio = noTemporario;
			}
			noTemporario.setAnterior(ponteiroFim);
			ponteiroFim = noTemporario;
			quantidade++;
		} else {
			System.out.println("Fila Cheia!");
		}	
	}
	
	//funciona como o desenfileirar de FilaDinamica
	@Override	
	public Object desenfileirarInicio() {
		Object dadoInicio = null;
		if(!estaVazia()) {
			dadoInicio = ponteiroInicio.getDado();
			ponteiroInicio = ponteiroInicio.getProximo();
			quantidade--;
			if (!estaVazia()) {
				ponteiroInicio.setAnterior(null);
			} else {
				ponteiroFim = null;
			}
		} else {
			System.out.println("Fila Vazia!");			
		}
		return dadoInicio;			
	}
	
	//funciona como o espiar de FilaDinamica
	@Override
	public Object espiarInicio() {
		Object dadoInicio = null;
		if(!estaVazia()) {
			dadoInicio = ponteiroInicio.getDado();
		} else {
			System.out.println("Fila Vazia!");			
		}
		return dadoInicio;
	}
	
	@Override
	public void enfileirarInicio(Object dado) {
		if(!estaCheia()) {
			NoDuplo noTemporario = new NoDuplo();
			noTemporario.setDado(dado);
			if (!estaVazia()) {
				ponteiroInicio.setAnterior(noTemporario);
			} else  {
				ponteiroFim = noTemporario;
			}
			noTemporario.setProximo(ponteiroInicio);
			ponteiroInicio = noTemporario;
			quantidade++;
		} else {
			System.out.println("Fila Cheia!");
		}	
	}
		
	@Override
	public Object desenfileirarFim() {
		Object dadoFim = null;
		if(!estaVazia()) {
			dadoFim = ponteiroFim.getDado();
			ponteiroFim = ponteiroFim.getAnterior();
			quantidade--;
			if (!estaVazia()) {
				ponteiroFim.setProximo(null);
			} else {
				ponteiroInicio = null;
			}
		} else {
			System.out.println("Fila Vazia!");			
		}
		return dadoFim;		
	}
			
	@Override
	public Object espiarFim() {
		Object dadoFim = null;
		if(!estaVazia()) {
			dadoFim = ponteiroFim.getDado();
		} else {
			System.out.println("Fila Vazia!");			
		}
		return dadoFim;
	}

	//funciona como o estaCheia de FilaDinamica	
	@Override
	public boolean estaCheia() {
		return (quantidade == tamanho);
	}

	//funciona como o estaVazia de FilaDinamica
	@Override
	public boolean estaVazia(){
		return (quantidade == 0);
	}
	
	//funciona como o imprimir de FilaDinamica
	@Override
	public String imprimir() {
		String resultado = "[";
		NoDuplo noAuxiliar = ponteiroInicio;
		for (int i = 0; i < quantidade; i++) {
			if (i == quantidade-1) {
				resultado += noAuxiliar.getDado();
			} else {
				resultado += noAuxiliar.getDado() + ",";
			}
			noAuxiliar = noAuxiliar.getProximo();
		}
		return resultado + "]";	
	}
}

File: fila/Fila Dinâmica/Kotlin/Fila Dupla Terminação/FilaDinamicaDuplaTerminacao.kt
class FilaDinamicaDuplaTerminacao(private val tamanho: Int = 10) : EnfileiravelDupla {

	private var ponteiroInicio: NoDuplo? = null
	private var ponteiroFim: NoDuplo? = null
	private var quantidade = 0
 
    override fun enfileirarInicio(dado: Any?) {
		if (!estaCheia()) {
			var noTemp = NoDuplo(dado)
			//noTemp.dado = dado
			noTemp.proximo = ponteiroInicio
			if (!estaVazia())
				ponteiroInicio?.anterior = noTemp
			else
				ponteiroFim = noTemp

			ponteiroInicio = noTemp
			quantidade = quantidade.inc()
		} else {
			println("Fila Cheia!")
		}
    }

	override fun enfileirarFim(dado: Any?) {
		if (!estaCheia()) {
			var noTemp = NoDuplo(dado)
			//noTemp.dado = dado
			noTemp.anterior = ponteiroFim
			if (!estaVazia())
				ponteiroFim?.proximo = noTemp
			else
				ponteiroInicio = noTemp

			ponteiroFim = noTemp
			quantidade = quantidade.inc()
		} else {
			println("Fila Cheia!")
		}
	}
	
    override fun desenfileirarFim(): Any? {
		var dadoAux: Any? = null
		if (!estaVazia()) {
			dadoAux = ponteiroFim?.dado
			ponteiroFim = ponteiroFim?.anterior
			quantidade = quantidade.dec()
			if (!estaVazia())
				ponteiroFim?.proximo = null
			else
				ponteiroInicio = null			
		} else {
			println("Fila Vazia!")
		}
		return dadoAux
	}

	override fun desenfileirarInicio(): Any? {
		var dadoAux: Any? = null
		if (!estaVazia()) {
			dadoAux = ponteiroInicio?.dado
			ponteiroInicio = ponteiroInicio?.proximo
			quantidade = quantidade.dec()
			if (!estaVazia())
				ponteiroInicio?.anterior = null
			else
				ponteiroFim = null			
		} else {
			println("Fila Vazia!")
		}
		return dadoAux
	}

	override fun tras(): Any? {
		var dadoAux: Any? = null
		if (!estaVazia())
			dadoAux = ponteiroFim?.dado
		else
			println("Fila Vazia!")

		return dadoAux
	}

	override fun frente(): Any? {
		var dadoAux: Any? = null
		if (!estaVazia())
			dadoAux = ponteiroInicio?.dado
		else
			println("Fila Vazia!")

		return dadoAux
	}

    override fun atualizarFim(dado: Any?) {
		if (!estaVazia())
			ponteiroFim?.dado = dado 
		else
			println("Fila Vazia!")
	}	

    override fun atualizarInicio(dado: Any?) {
		if (!estaVazia())
			ponteiroInicio?.dado = dado 
		else
			println("Fila Vazia!")
	}	

	override fun estaCheia(): Boolean {
		return quantidade == tamanho
	}
	
	override fun estaVazia(): Boolean {
		return quantidade == 0
	}
	
    override fun imprimirTrasPraFrente(): String {
		var ponteiroAuxiliar = ponteiroFim
		var resultado = "["
		for (i in 0 until quantidade) {
			resultado += ponteiroAuxiliar?.dado	
			if (i != quantidade-1)
				resultado += ","			
			
			ponteiroAuxiliar = ponteiroAuxiliar?.anterior
		}
		return "$resultado]"
    }
    
	override fun imprimirFrentePraTras(): String {
		var ponteiroAuxiliar = ponteiroInicio
		var resultado = "["
		for (i in 0 until quantidade) {
			resultado += ponteiroAuxiliar?.dado
			if (i != quantidade-1)
				resultado += ","

			ponteiroAuxiliar = ponteiroAuxiliar?.proximo
		}
		return "$resultado]"
	}
}